packages/flutter/lib/src/cupertino/text_selection_toolbar.dart

```diff
@@ -510,49 +520,77 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
   Widget build(BuildContext context) {
     return widget.toolbarBuilder(context, widget.anchor, widget.isAbove, FadeTransition(
       opacity: _controller,
-      child: _CupertinoTextSelectionToolbarItems(
-        page: _page,
-        backButton: CupertinoTextSelectionToolbarButton.text(
-          onPressed: _handlePreviousPage,
-          text: '◀',
-        ),
-        dividerWidth: 1.0 / MediaQuery.devicePixelRatioOf(context),
-        nextButton: CupertinoTextSelectionToolbarButton.text(
-          onPressed: _handleNextPage,
-          text: '▶',
-        ),
-        nextButtonDisabled: const CupertinoTextSelectionToolbarButton.text(
-          text: '▶',
+      child: AnimatedSize(
+        duration: _kToolbarTransitionDuration,
+        curve: Curves.decelerate,
+        child: GestureDetector(
+          onHorizontalDragEnd: _onHorizontalDragEnd,
+          child: _CupertinoTextSelectionToolbarItems(
+            controller: _toolbarItemsController,
+            page: _page,
+            backButton: CupertinoTextSelectionToolbarButton(
+              onPressed: _handlePreviousPage,
+              child: Icon(
+                CupertinoIcons.chevron_left,
+                color: _kToolbarTextColor.resolveFrom(context),
+                size: _kToolbarChevronSize,
+              ),
+            ),
+            dividerColor: _kToolbarDividerColor.resolveFrom(context),
+            dividerWidth: 1.0 / MediaQuery.devicePixelRatioOf(context),
+            nextButton: CupertinoTextSelectionToolbarButton(
+              onPressed: _handleNextPage,
+              child: Icon(
+                CupertinoIcons.chevron_right,
+                color: _kToolbarTextColor.resolveFrom(context),
+                size: _kToolbarChevronSize,
+              ),
+            ),
+            children: widget.children,
+          ),
         ),
-        children: widget.children,
       ),
     ));
   }
 }
 
+// This doesn't really control the toolbar, but has updated values to check if
+// there are previous/next pages. This wasn't needed before because the toolbar
+// correctly shows/hides the previous/next buttons automatically, but now the
+// horizontal drag gesture callback has to check whether it's possible to
+// navigate or not, making this necessary.
+class _CupertinoTextSelectionToolbarItemsController {
```

Comment:
I think controller classes are normally used in APIs for others widgets to change the state of a stateful widget. Here we're operating on a render object so this should probably be done using individual fields e.g.

```
void updateRenderObject(BuildContext context, _RenderCupertinoTextSelectionToolbarItems renderObject) {
    renderObject
      ..page = page
      ..hasPreviousPage = hasPreviousPage
      ..hasNextPage = hasNextPage
      ..dividerColor = dividerColor
```

Then in the `RenderBox` class you'd have a setter and a getter just like with `dividerColor`.

Reply:
The problem with this approach is that, as far as I understand it, I wouldn't be able to update `hasNextPage` and `hasPreviousPage` in `_RenderCupertinoTextSelectionToolbarItems`, and have it available in its parent (`_CupertinoTextSelectionToolbarContentState`) when the user does a horizontal drag gesture on the whole toolbar.

This is where I update the values:
https://github.com/flutter/flutter/blob/6762521feb2821287b66bbd315cfe3b2fba56857/packages/flutter/lib/src/cupertino/text_selection_toolbar.dart#L944-L947

And this is where I use them:
https://github.com/flutter/flutter/blob/6762521feb2821287b66bbd315cfe3b2fba56857/packages/flutter/lib/src/cupertino/text_selection_toolbar.dart#L461-L475

Reply:
If the `hasPreviousPage` / `hasNextPage` fields are only updated by `_RenderCupertinoTextSelectionToolbarItems` (which makes sense given this will depend on the incoming constraints for the RenderBox), perhaps one approach would be to:
- keep them as `bool` values without getter/setter
- from the widget state, in the gesture handler, do something like:
```
final RenderBox? renderBox = context.findRenderObject() as RenderBox?;
if (renderBox is _RenderCupertinoTextSelectionToolbarItems) {
  if (renderBox.hasPreviousPage...)

```

Reply:
That makes sense and it's way cleaner, thank you! I will see if I can remove the controller until this weekend.

Reply:
Just fixed this and all tests passed, should be OK now 😁